File: simulation/core/rewards.js
/**
 * Reward Distribution Module
 * Handles participation-normalized reward calculations
 */

import CONFIG from '../config.js';
import { calculateEmission } from './emission.js';

/**
 * Smoothed total effectiveness tracker
 * Uses exponential moving average to reduce volatility
 */
export class SmoothedEffectiveness {
    constructor() {
        this.value = 0;
        this.alpha = CONFIG.smoothing.alpha;
        this.initialized = false;
    }

    /**
     * Update with new total effectiveness
     * @param {number} totalEffectiveness - Current total effectiveness
     * @returns {number} Smoothed value
     */
    update(totalEffectiveness) {
        if (!this.initialized) {
            this.value = totalEffectiveness;
            this.initialized = true;
        } else {
            this.value = this.alpha * totalEffectiveness + (1 - this.alpha) * this.value;
        }
        return this.value;
    }

    /**
     * Get current smoothed value
     * @returns {number}
     */
    getValue() {
        return this.value;
    }

    /**
     * Reset the smoother
     */
    reset() {
        this.value = 0;
        this.initialized = false;
    }
}

/**
 * Calculate reward for a single node
 * Reward_i = Em_epoch * (Eff_i / S_t)
 * 
 * @param {number} nodeEffectiveness - Node's effectiveness weight
 * @param {number} smoothedTotalEffectiveness - Smoothed total of all effectiveness
 * @param {number} epochEmission - Total tokens emitted this epoch
 * @returns {number} Tokens rewarded to this node
 */
export function calculateNodeReward(nodeEffectiveness, smoothedTotalEffectiveness, epochEmission) {
    if (smoothedTotalEffectiveness === 0) {
        return 0;
    }
    return epochEmission * (nodeEffectiveness / smoothedTotalEffectiveness);
}

/**
 * Calculate rewards for all nodes in a network
 * 
 * @param {Array<{id: string, effectiveness: number}>} nodes - Array of node data
 * @param {number} daysSinceGenesis - Current day
 * @param {SmoothedEffectiveness} smoother - Smoothed effectiveness tracker
 * @returns {Object} Reward distribution data
 */
export function calculateNetworkRewards(nodes, daysSinceGenesis, smoother) {
    const totalEffectiveness = nodes.reduce((sum, node) => sum + node.effectiveness, 0);
    const smoothedTotal = smoother.update(totalEffectiveness);
    const epochEmission = calculateEmission(daysSinceGenesis);

    const rewards = nodes.map(node => ({
        id: node.id,
        effectiveness: node.effectiveness,
        reward: calculateNodeReward(node.effectiveness, smoothedTotal, epochEmission),
        share: totalEffectiveness > 0 ? node.effectiveness / totalEffectiveness : 0,
    }));

    return {
        day: daysSinceGenesis,
        epochEmission,
        totalEffectiveness,
        smoothedTotalEffectiveness: smoothedTotal,
        nodeCount: nodes.length,
        rewards,
        giniCoefficient: calculateGini(rewards.map(r => r.reward)),
    };
}

/**
 * Calculate Gini coefficient for inequality measurement
 * 
 * @param {number[]} values - Array of values (e.g., rewards)
 * @returns {number} Gini coefficient (0 = perfect equality, 1 = perfect inequality)
 */
export function calculateGini(values) {
    if (values.length === 0) return 0;

    const n = values.length;
    const sorted = [...values].sort((a, b) => a - b);
    const sum = sorted.reduce((a, b) => a + b, 0);

    if (sum === 0) return 0;

    let cumulativeSum = 0;
    let weightedSum = 0;

    for (let i = 0; i < n; i++) {
        cumulativeSum += sorted[i];
        weightedSum += (i + 1) * sorted[i];
    }

    return (2 * weightedSum) / (n * sum) - (n + 1) / n;
}

/**
 * Simulate reward dilution when new nodes join
 * 
 * @param {number} existingNodes - Number of existing full-effectiveness nodes
 * @param {number} newNodes - Number of new nodes joining
 * @param {number} newNodeEffectiveness - Effectiveness of new nodes (0-1)
 * @param {number} epochEmission - Tokens emitted per epoch
 * @returns {Object} Dilution analysis
 */
export function analyzeRewardDilution(existingNodes, newNodes, newNodeEffectiveness, epochEmission) {
    // Before new nodes
    const beforeTotalEff = existingNodes * 1.0; // Assume full effectiveness
    const beforeRewardPerNode = epochEmission / existingNodes;

    // After new nodes
    const afterTotalEff = beforeTotalEff + (newNodes * newNodeEffectiveness);
    const afterRewardExisting = epochEmission * (1.0 / afterTotalEff);
    const afterRewardNew = epochEmission * (newNodeEffectiveness / afterTotalEff);

    return {
        before: {
            totalEffectiveness: beforeTotalEff,
            rewardPerNode: beforeRewardPerNode,
        },
        after: {
            totalEffectiveness: afterTotalEff,
            rewardPerExistingNode: afterRewardExisting,
            rewardPerNewNode: afterRewardNew,
        },
        dilutionPercent: ((beforeRewardPerNode - afterRewardExisting) / beforeRewardPerNode) * 100,
    };
}

export default {
    SmoothedEffectiveness,
    calculateNodeReward,
    calculateNetworkRewards,
    calculateGini,
    analyzeRewardDilution,
};

File: simulation_py/rewards.py
"""
Reward Distribution Module
Handles participation-normalized reward calculations
"""

from config import CONFIG
from emission import calculate_emission


class SmoothedEffectiveness:
    """
    Smoothed total effectiveness tracker
    Uses exponential moving average to reduce volatility
    """
    
    def __init__(self):
        self.value = 0
        self.alpha = CONFIG["smoothing"]["alpha"]
        self.initialized = False
    
    def update(self, total_effectiveness: float) -> float:
        """Update with new total effectiveness"""
        if not self.initialized:
            self.value = total_effectiveness
            self.initialized = True
        else:
            self.value = self.alpha * total_effectiveness + (1 - self.alpha) * self.value
        return self.value
    
    def get_value(self) -> float:
        """Get current smoothed value"""
        return self.value
    
    def reset(self):
        """Reset the smoother"""
        self.value = 0
        self.initialized = False


def calculate_node_reward(node_effectiveness: float, smoothed_total: float, 
                          epoch_emission: float) -> float:
    """
    Calculate reward for a single node
    Reward_i = Em_epoch * (Eff_i / S_t)
    
    Args:
        node_effectiveness: Node's effectiveness weight
        smoothed_total: Smoothed total of all effectiveness
        epoch_emission: Total tokens emitted this epoch
        
    Returns:
        Tokens rewarded to this node
    """
    if smoothed_total == 0:
        return 0
    return epoch_emission * (node_effectiveness / smoothed_total)


def calculate_network_rewards(nodes: list, days_since_genesis: int, 
                              smoother: SmoothedEffectiveness) -> dict:
    """
    Calculate rewards for all nodes in a network
    
    Args:
        nodes: List of node dicts with id and effectiveness
        days_since_genesis: Current day
        smoother: Smoothed effectiveness tracker
        
    Returns:
        Reward distribution data
    """
    total_effectiveness = sum(node["effectiveness"] for node in nodes)
    smoothed_total = smoother.update(total_effectiveness)
    epoch_emission = calculate_emission(days_since_genesis)
    
    rewards = []
    for node in nodes:
        reward = calculate_node_reward(node["effectiveness"], smoothed_total, epoch_emission)
        share = node["effectiveness"] / total_effectiveness if total_effectiveness > 0 else 0
        rewards.append({
            "id": node["id"],
            "effectiveness": node["effectiveness"],
            "reward": reward,
            "share": share,
        })
    
    return {
        "day": days_since_genesis,
        "epoch_emission": epoch_emission,
        "total_effectiveness": total_effectiveness,
        "smoothed_total_effectiveness": smoothed_total,
        "node_count": len(nodes),
        "rewards": rewards,
        "gini_coefficient": calculate_gini([r["reward"] for r in rewards]),
    }


def calculate_gini(values: list) -> float:
    """
    Calculate Gini coefficient for inequality measurement
    
    Args:
        values: List of values (e.g., rewards)
        
    Returns:
        Gini coefficient (0 = perfect equality, 1 = perfect inequality)
    """
    if not values:
        return 0
    
    n = len(values)
    sorted_values = sorted(values)
    total = sum(sorted_values)
    
    if total == 0:
        return 0
    
    weighted_sum = sum((i + 1) * v for i, v in enumerate(sorted_values))
    
    return (2 * weighted_sum) / (n * total) - (n + 1) / n


def analyze_reward_dilution(existing_nodes: int, new_nodes: int, 
                            new_node_effectiveness: float, epoch_emission: float) -> dict:
    """
    Simulate reward dilution when new nodes join
    
    Args:
        existing_nodes: Number of existing full-effectiveness nodes
        new_nodes: Number of new nodes joining
        new_node_effectiveness: Effectiveness of new nodes (0-1)
        epoch_emission: Tokens emitted per epoch
        
    Returns:
        Dilution analysis
    """
    # Before new nodes
    before_total_eff = existing_nodes * 1.0  # Assume full effectiveness
    before_reward_per_node = epoch_emission / existing_nodes
    
    # After new nodes
    after_total_eff = before_total_eff + (new_nodes * new_node_effectiveness)
    after_reward_existing = epoch_emission * (1.0 / after_total_eff)
    after_reward_new = epoch_emission * (new_node_effectiveness / after_total_eff)
    
    dilution = ((before_reward_per_node - after_reward_existing) / before_reward_per_node) * 100
    
    return {
        "before": {
            "total_effectiveness": before_total_eff,
            "reward_per_node": before_reward_per_node,
        },
        "after": {
            "total_effectiveness": after_total_eff,
            "reward_per_existing_node": after_reward_existing,
            "reward_per_new_node": after_reward_new,
        },
        "dilution_percent": dilution,
    }
